src/language-js/postprocess.js

```diff
@@ -5,8 +5,42 @@ const {
   getNextNonSpaceNonCommentCharacter
 } = require("../common/util");
 const { composeLoc, locEnd } = require("./loc");
+const { isTypeCastComment } = require("./comments");
 
 function postprocess(ast, options) {
+  // Keep Babel's non-standard ParenthesizedExpression nodes only if they have Closure-style type cast comments.
+  if (options.parser !== "typescript" && options.parser !== "flow") {
+    const startOffsetsOfTypeCastedNodes = new Set();
+
+    // Comments might be attached not directly to ParenthesizedExpression but to its ancestor.
+    // E.g.: /** @type {Foo} */ (foo).bar();
+    // Let's use the fact that those ancestors and ParenthesizedExpression have the same start offset.
+
+    visitNode(ast, node => {
+      if (
+        node.leadingComments &&
+        node.leadingComments.some(isTypeCastComment)
+      ) {
+        startOffsetsOfTypeCastedNodes.add(node.start);
+      }
+    });
+
+    visitNode(ast, node => {
+      if (
+        node.type === "ParenthesizedExpression" &&
+        !startOffsetsOfTypeCastedNodes.has(node.start)
+      ) {
+        const { expression } = node;
+        if (!expression.extra) {
+          expression.extra = {};
+        }
+        expression.extra.parenthesized = true;
+        expression.extra.parenStart = node.start;
+        return expression;
```

Comment:
What about open a issue in `babel` repo about it? I think they can improve, we have pretty good reasons for this

Reply:
I don't see what you mean. Creating this `extra` stuff is Babel's default behavior, but when `createParenthesizedExpressions` is on, Babel doesn't create it. However, some other code in Prettier relied on `extra.parenthesized `, so I couldn't just enable `createParenthesizedExpressions` without recreating `extra`.